packages/teleport-generator-core/src/validator/utils.ts

```diff
@@ -0,0 +1,186 @@
+import {
+  traverseNodes,
+  traverseElements,
+} from '@teleporthq/teleport-generator-shared/lib/utils/uidl-utils'
+
+import {
+  ProjectUIDL,
+  UIDLElement,
+  ComponentUIDL,
+} from '@teleporthq/teleport-generator-shared/lib/typings/uidl'
+
+// Prop definitions and state definitions should have different keys
+export const checkForDuplicateDefinitions = (input: ComponentUIDL) => {
+  const props = Object.keys(input.propDefinitions || {})
+  const states = Object.keys(input.stateDefinitions || {})
+
+  props
+    .filter((x) => states.includes(x))
+    .map((duplicate) =>
+      console.warn(
+        `\n"${duplicate}" is defined both as a prop and as a state. If you are using VUE Code Generators this can cause bad behavior.`
+      )
+    )
+}
+
+// In "repeat" node:
+// If index is used, "useIndex" must be declared in "meta"
+// If custom local variable is used, it's name must be specified inside "meta" as "iteratorName"
+export const checkForLocalVariables = (input: ComponentUIDL) => {
+  const errors = []
+
+  traverseNodes(input.node, (node) => {
+    if (node.type === 'repeat') {
+      traverseNodes(node, (childNode) => {
+        if (childNode.type === 'dynamic' && childNode.content.referenceType === 'local') {
+          if (
+            childNode.content.id &&
+            node.content.meta.iteratorName &&
+            childNode.content.id !== node.content.meta.iteratorName
+          ) {
+            const errorMsg = `\n"${
+              childNode.content.id
+            }" is used in the "repeat" structure but the iterator name has this value: "${
+              node.content.meta.iteratorName
+            }"`
+            errors.push(errorMsg)
+          }
+          if (childNode.content.id && !node.content.meta.useIndex) {
+            const errorMsg = `\nIndex variable is used but the "useIndex" meta information is false.`
+            errors.push(errorMsg)
+          }
+        }
+      })
+    }
+  })
+  return { errors }
+}
+
+// All referenced props and states should be previously defined in the
+// "propDefinitions" and "stateDefinitions" sections
+// If props or states are defined and not used, a warning witll be displayed
+export const checkDynamicDefinitions = (input: any) => {
+  const definedKeys = Object.keys(input.propDefinitions || {}).concat(
+    Object.keys(input.stateDefinitions || {})
+  )
+  const usedKeys = []
+  const errors = []
+
+  traverseNodes(input.node, (node) => {
+    if (
+      node.type === 'dynamic' &&
+      (node.content.referenceType === 'prop' || node.content.referenceType === 'state')
```

Comment:
I think you need to check separately here, if refType is prop, you should explicitly check the key exists in `propDefinitions`. similarly, when the refType is state, you should check in `stateDefinitions`